File: src/main/java/org/oewntk/model/Pronunciation.java
/*
 * Copyright (c) 2021-2021. Bernard Bou.
 */

package org.oewntk.model;

import java.io.Serializable;
import java.util.Objects;

/**
 * Pronunciation
 */
public class Pronunciation implements Serializable
{
	private final String value;

	private final String variety;

	public static Pronunciation ipa(final String value)
	{
		return new Pronunciation(value, null);
	}

	public static Pronunciation ipa(final String value, final String variety)
	{
		return new Pronunciation(value, variety);
	}

	/**
	 * Constructor
	 *
	 * @param value   value in IPA
	 * @param variety variety
	 */
	public Pronunciation(final String value, final String variety)
	{
		this.value = value;
		this.variety = variety;
	}

	/**
	 * Get value in IPA
	 *
	 * @return value in IPA
	 */
	public String getValue()
	{
		return value;
	}

	/**
	 * Get variety
	 *
	 * @return variety
	 */
	public String getVariety()
	{
		return variety;
	}

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Pronunciation that = (Pronunciation) o;
		return Objects.equals(value, that.value) && Objects.equals(variety, that.variety);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(value, variety);
	}

	// stringify

	@Override
	public String toString()
	{
		String value = "/" + this.value + '/';
		if (variety != null)
		{
			return String.format("[%s] %s", variety, value);
		}
		return value;
	}
}

File: src/main/java/org/oewntk/model/Pronunciation.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import java.io.Serializable
import java.util.*

/**
 * Pronunciation
 *
 * @property value   value in IPA
 * @property variety variety
 */
@kotlinx.serialization.Serializable
data class Pronunciation(
    val value: PronunciationValue,
    val variety: PronunciationVariety?,
) : Comparable<Pronunciation>, Serializable {

    // identify

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as Pronunciation
        return value == that.value && variety == that.variety
    }

    override fun hashCode(): Int {
        return Objects.hash(value, variety)
    }

    // stringify

    override fun toString(): String {
        return if (variety != null) "[$variety] /$value/" else "/$value/"
    }

    // order

    override fun compareTo(other: Pronunciation): Int {
        return comparator.compare(this, other)
    }

    companion object {

        @Transient
        val comparator: Comparator<Pronunciation> = compareBy(Pronunciation::variety).thenBy(Pronunciation::value)

        @Suppress("unused")
        @Transient
        val comparatorNull: Comparator<Pronunciation?> = nullsFirst(comparator)

        fun ipa(value: PronunciationValue): Pronunciation {
            return Pronunciation(value, null)
        }

        fun ipa(value: PronunciationValue, variety: PronunciationVariety?): Pronunciation {
            return Pronunciation(value, variety)
        }
    }
}
